### PR TITLE
use '-' as default separator and support of compatibility mode

### PR DESCRIPTION
**Pull request by glours**
Change the default separator used between project name and resources to `-`. This is especially needed for container names to avoid issue with hostname support.

Also add detection of the `COMPOSE_COMPATIBILITY` env variable to support the compatibility usage of the `_` separator

Fix https://github.com/docker/compose/issues/9618

### CODE REVIEW

File: cli/options_test.go [hunk 1]
```diff
@@ -237,3 +237,29 @@ func TestEnvMap(t *testing.T) {
...
+		assert.NilError(t, err)
+		p, err := ProjectFromOptions(opts)
+		assert.NilError(t, err)
+		assert.Equal(t, p.Networks["test-network"].Name, "my-project-test-network")
+
+	})
+
+	t.Run("With compatibility separator", func(t *testing.T) {
+		os.Setenv("COMPOSE_COMPATIBILITY", "true") //nolint:errcheck
+		defer os.Unsetenv("COMPOSE_COMPATIBILITY") //nolint:errcheck
```
Comment (milas) [inline cli/options_test.go:255]:
```suggestion
		t.Setenv("COMPOSE_COMPATIBILITY", "true")
```
It'll auto unset for you and ensure that the test doesn't get run in parallel by accident 🙂

**Review by milas — APPROVED** [review]
LGTM

File: cli/options_test.go (hunk 1 above)
Comment (laurazard) [reply cli/options_test.go:255]: oooh this is nice to know!

Comment (glours) [reply cli/options_test.go:255]: Good catch @milas, I always forgot about that one!!!

**Review by milas — APPROVED** [review]

File: cli/options.go [hunk 2]
```diff
@@ -384,6 +385,15 @@ func withConvertWindowsPaths(options *ProjectOptions) func(*loader.Options) {
 	}
 }
 
+// withSeparator defines loader.Options separator used to define resource names
+func withSeparator(options *ProjectOptions) func(*loader.Options) {
+	return func(o *loader.Options) {
+		if utils.StringToBool(options.Environment["COMPOSE_COMPATIBILITY"]) {
+			o.Separator = loader.CompatibilitySeparator
+		}
```
Comment (ulyssessouza) [inline cli/options.go:393]: Please note that this only fixes https://github.com/docker/compose/issues/9618 when passing an environment variable. In the case of Docker Compose getting a `--compatibility` from the command line I'm not sure that will work  (didn't test though).

Given this: https://github.com/docker/compose/blob/v2/cmd/compose/compose.go#L162-L169

Looks like compose just evaluates the command line option after this code is already evaluated.
One solution would be checking the command line option before and setting the EnvVar before calling `ProjectFromOptions` in the Docker Compose repo.

**Review by ulyssessouza — CHANGES_REQUESTED** [review]

**Review by ulyssessouza — APPROVED** [review]

**Comment by logopk** [comment]
Stupid breaking change - duplicates all my networks and returns  conflicting ip ranges.
`        "failed to create network step-ca_t-default: Error response from daemon: Pool overlaps with other one on this address space" `
Quite a hassle to get this fixed!

**Comment by MrNaif2018** [comment]
Compose v2 automatically changed container names for me when just using `docker compose up` command. When upgrading to 2.8.0 it did create new volumes and network, but containers used old ones, so it requires a full restart, also volumes data isn't migrated automatically but probably should be

**Comment by glours** [comment]
If you want to keep the previous behaviour, you use the `--compatibility` flag
```shell
docker compose --compatibility up -d
[+] Running 5/5
 ⠿ Network react-express-mongodb_react-express  Created                                                                                                                                                    0.1s
 ⠿ Network react-express-mongodb_express-mongo  Created                                                                                                                                                    0.0s
 ⠿ Container mongo                              Started                                                                                                                                                    4.8s
 ⠿ Container backend                            Started                                                                                                                                                    5.1s
 ⠿ Container frontend                           Started
```

**Comment by logopk** [comment]
???

i need to ediit ALL of my scripts?

**Comment by MrNaif2018** [comment]
I think the issue here is that --compatibility flag is for compose v1 -> v2 migration, but breaking change (which should have probably been in v2.0) was introduced in a minor release 2.8.0
We'll try to adapt it and move docker volumes ourselves, docker compose v2 is stable to be used in production, right?

**Comment by ulyssessouza** [comment]
You can also use an environment variable for that:
`export COMPOSE_COMPATIBILITY=true`

Like this you don't have to use `--compatibility` on every call on your scripts

**Comment by glours** [comment]
@MrNaif2018 yes you can use it production, End-Of-Life of Compose V1 was announced months ago

**Comment by logopk** [comment]
I still do think this is absurd. All running environments need to be rebuilt - even with recreation of all volumes.
Plus all connections to external networks/volumes" need to be rewritten!

I do understand that the environment variable fixes this. But what other side effects has this, as I'm on v2 for a long time and I don't need compatibility with v1!

Me don't like this. It's a minor version release! Please reconsider and make this an opt in.

**Comment by glours** [comment]
@logopk I will rollback the changes and do releases Compose and compose-go

**Comment by glours** [comment]
@MrNaif2018 FYI we're rolling back the changes, do be to fast in your volumes migration!

**Comment by glours** [comment]
A new release of [compose-go available](https://github.com/compose-spec/compose-go/releases/tag/v1.4.0) without the breaking changes, a version of Docker Compose should follow soon
cc @logopk @MrNaif2018 @AkihiroSuda

**Comment by MrNaif2018** [comment]
@glours Oh I see. I didn't start implementing the migration script luckily. What should my next steps be? I mean I am migrating all deployments from compose v1 to v2 and finished it, but in the meantime when I didn't yet release the changes 2.8.0 release occurred. Ideally in our scripts I would install the latest version of compose v2: https://github.com/bitcartcc/bitcart-docker/blob/dc7de21a1826ede14318701a668956935c64a5d8/helpers.sh#L155-L174

What should my further actions be? Should I prepare that those changes will return back at a later point in compose v2, or they won't? Or they will return but with auto-migration from both compose 1.x and 2.x to the new version like it's done for container re-creation? I'm fine either way, ideally I would migrate all users to v2+ and then no weird bug reports would occur (:

**Comment by glours** [comment]
@MrNaif2018 just use `v2.9.0` of Compose which was just released.
The value of those changes regarding the impact on existing Compose stacks running is to low, so we won't reintroduce them.

**Comment by logopk** [comment]
Thank you for the quick response. Much appreciated!

**Comment by jowax** [comment]
@glours I'm still getting the `-` separator in docker-compose 2.9.0

Compose file:
```sh
version: '3.8'

services:
  app:
    build:
      context: .
      dockerfile: Dockerfile
```

Version 2.9.0
```sh
~/projects/foo ❯ ./docker-compose-2.9.0 --version                                                                                                                                                                                                
Docker Compose version v2.9.0
~/projects/foo ❮ ./docker-compose-2.9.0 --project-name foo_devcontainer -f docker-compose.yml build                                                                                                                                              
[+] Building 0.2s (5/5) FINISHED                                                                                                                                                                                                                                                          
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                 0.0s
 => => transferring dockerfile: 31B                                                                                                                                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                    0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                                      0.0s
 => [internal] load metadata for docker.io/library/ubuntu:20.04                                                                                                                                                                                                                      0.2s
 => CACHED [1/1] FROM docker.io/library/ubuntu:20.04@sha256:af5efa9c28de78b754777af9b4d850112cad01899a5d37d2617bb94dc63a49aa                                                                                                                                                         0.0s
 => exporting to image                                                                                                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                                                                                                              0.0s
 => => writing image sha256:b6139e4d50625d70c36fc0232ae66d887f4a58115d9579a989f4490f0a65cf28                                                                                                                                                                                         0.0s
 => => naming to docker.io/library/foo_devcontainer-app                                                                                                                                                                                                                              0.0s
```

Version 2.8.0
```sh
~/projects/foo ❯ ./docker-compose-2.8.0 --version                                                                                                                                                                                               
Docker Compose version v2.8.0
~/projects/foo ❮ ./docker-compose-2.8.0 --project-name foo_devcontainer -f docker-compose.yml build                                                                                                                                              
[+] Building 0.5s (5/5) FINISHED                                                                                                                                                                                                                                                          
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                 0.0s
 => => transferring dockerfile: 31B                                                                                                                                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                    0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                                      0.0s
 => [internal] load metadata for docker.io/library/ubuntu:20.04                                                                                                                                                                                                                      0.4s
 => CACHED [1/1] FROM docker.io/library/ubuntu:20.04@sha256:af5efa9c28de78b754777af9b4d850112cad01899a5d37d2617bb94dc63a49aa                                                                                                                                                         0.0s
 => exporting to image                                                                                                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                                                                                                              0.0s
 => => writing image sha256:b6139e4d50625d70c36fc0232ae66d887f4a58115d9579a989f4490f0a65cf28                                                                                                                                                                                         0.0s
 => => naming to docker.io/library/foo_devcontainer-app                                                                                                                                                                                                                              0.0s
```

Version 2.7.0
```sh
~/projects/foo ❯ ./docker-compose-2.7.0 --version                                                                                                                                                                                                
Docker Compose version v2.7.0
~/projects/foo ❯ ./docker-compose-2.7.0 --project-name foo_devcontainer -f docker-compose.yml build                                                                                                                                              
[+] Building 0.6s (5/5) FINISHED                                                                                                                                                                                                                                                          
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                 0.0s
 => => transferring dockerfile: 31B                                                                                                                                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                    0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                                      0.0s
 => [internal] load metadata for docker.io/library/ubuntu:20.04                                                                                                                                                                                                                      0.5s
 => CACHED [1/1] FROM docker.io/library/ubuntu:20.04@sha256:af5efa9c28de78b754777af9b4d850112cad01899a5d37d2617bb94dc63a49aa                                                                                                                                                         0.0s
 => exporting to image                                                                                                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                                                                                                              0.0s
 => => writing image sha256:b6139e4d50625d70c36fc0232ae66d887f4a58115d9579a989f4490f0a65cf28                                                                                                                                                                                         0.0s
 => => naming to docker.io/library/foo_devcontainer_app                                                                                                                                                                                                                              0.0s
```

**Comment by jamshid** [comment]
@glours  wrote:
> The value of those changes regarding the impact on existing Compose stacks running is to low, so we won't reintroduce them.

Eh ok but the network name can end up in a fully qualified dns name:
```
# nslookup 172.23.0.15
15.0.23.172.in-addr.arpa	name = myproject-myservice-2.myproject_default.
```
and at least python is very strict about underscores, leading to `idna.core.InvalidCodepoint: Codepoint U+005F` errors e.g.:
```
          File "/tmp/build/results/usr/local/lib/python3.9/dist-packages/twisted/python/urlpath.py", line 265, in __str__
          File "/tmp/build/results/usr/local/lib/python3.9/dist-packages/hyperlink/_url.py", line 1668, in to_uri
          File "/tmp/build/results/usr/local/lib/python3.9/dist-packages/idna/core.py", line 360, in encode
          File "/tmp/build/results/usr/local/lib/python3.9/dist-packages/idna/core.py", line 258, in alabel
          File "/tmp/build/results/usr/local/lib/python3.9/dist-packages/idna/core.py", line 297, in ulabel
          File "/tmp/build/results/usr/local/lib/python3.9/dist-packages/idna/core.py", line 250, in check_label
        idna.core.InvalidCodepoint: Codepoint U+005F at position 10 of 'myproject_default' not allowed
```
Workaround: specify the network name in docker-compose.yml:
```
networks:
  default:
    name: myproject-default
    driver: bridge
    attachable: true
```

**Comment by AlexMikhalev** [comment]
You don't realise how much work such changes generate for projects. Mine is a medium-complexity project https://github.com/applied-knowledge-systems/the-pattern/ where deployment is now broken in multiple places. I spent time two years ago making all deployments docker and docker-compose based. Going forward, I will ditch docker-compose and only use docker for builds. Try to reach a point where benchmarks work: https://reference-architecture.ai/docs/bert-qa-benchmarking/ and track the effort required to update all docker-compose files.
Looking forward to your PRs.

**Pull request merged**